mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,644 @@
+            opt_state,
+            buffer_state,
+            params,
+            key,
+        ) = learner_state
+        new_key, interact_key, train_key = jax.random.split(key, 3)
+
+        # Select actions, step env and store transitions
+        action_selection_state = ActionSelectionState(
+            params.online, hidden_state, time_steps, interact_key
+        )
+        interaction_state = InteractionState(
+            action_selection_state, env_state, buffer_state, obs, done
+        )
+        final_interaction_state, metrics = scanned_interact(interaction_state)
+
+        # Sample and learn
+        train_state = TrainState(
+            final_interaction_state.buffer_state, params, opt_state, train_steps, train_key
+        )
+        final_train_state, losses = scanned_train(train_state)
+
+        next_learner_state = LearnerState(
+            final_interaction_state.obs,
+            final_interaction_state.done,
+            final_interaction_state.action_selection_state.hidden_state,
+            final_interaction_state.env_state,
+            final_interaction_state.action_selection_state.time_steps,
+            final_train_state.train_steps,
+            final_train_state.opt_state,
+            final_interaction_state.buffer_state,
+            final_train_state.params,
+            new_key,
+        )
+
+        return next_learner_state, (metrics, losses)
+
+    pmaped_update_step = jax.pmap(
+        jax.vmap(
+            lambda state: lax.scan(update_step, state, None, length=cfg.system.scan_steps),
+            axis_name="batch",
+        ),
+        axis_name="device",
+        donate_argnums=0,
+    )
+
+    return pmaped_update_step  # type:ignore
+
+
+def run_experiment(cfg: DictConfig) -> float:
+    # Add runtime variables to config
+    cfg.arch.n_devices = len(jax.devices())
+
+    # Number of env steps before evaluating/logging.
+    steps_per_rollout = int(cfg.system.total_timesteps // cfg.arch.num_evaluation)
+    # Multiplier for a single env/learn step in an anakin system
+    anakin_steps = cfg.arch.n_devices * cfg.system.update_batch_size
+    # Number of env steps in one anakin style update.
+    anakin_act_steps = anakin_steps * cfg.arch.num_envs * cfg.system.rollout_length
+    # Number of steps to do in the scanned update method (how many anakin steps).
+    cfg.system.scan_steps = int(steps_per_rollout / anakin_act_steps)
+
+    pprint(OmegaConf.to_container(cfg, resolve=True))
+
+    # Initialise system and make learning/evaluation functions
+    (env, eval_env), q_net, opt, rb, learner_state, logger, key = init(cfg)
+    update = make_update_fns(cfg, env, q_net, opt, rb)
+
+    cfg.system.num_agents = env.num_agents
+
+    key, eval_key = jax.random.split(key)
+    evaluator, absolute_metric_evaluator = make_eval_fns(
+        eval_env=eval_env,
+        network_apply_fn=q_net.apply,
+        config=cfg,
+        use_recurrent_net=True,
+        scanned_rnn=ScannedRNN(),
```

Comment:
Just adding a general note here, the `ScannedRNN` can take in a hidden_state_dim argument now which will help a lot with initialising hiddenstates. Can we please do this here and elsewhere where the `ScannedRNN` is used?

Reply:
Perfect, done! I want to make a PR for changing the ScannedRNN to automatically use its hidden_state_dim variabl in the initialise_carry function always, will that make sense for the other systems as well?